tiled/client/xarray.py

```diff
@@ -1,540 +1,94 @@
-        else:
-            try:
-                content = self.context.get_json(
-                    self.uri, params={"fields": "structure.macro", **self._params}
-                )
-                macro = content["data"]["attributes"]["structure"]["macro"]
-                variables = [*macro["data_vars"], *macro["coords"]]
-            except Exception:
-                # Do not print messy traceback from thread. Just fail silently.
-                return []
-        return variables
+        return list(self)
 
     def download(self):
         super().download()
-        self._ipython_key_completions_()
+        list(self)
         self.read().compute()
 
-    @property
-    def data_vars(self):
-        structure = self.structure().macro
-        _coords_clients, coords = self._build_coords_clients(structure)
-        # Turn off wide table optimization so that we return normal
-        # [Dask]DataArrayClients and not internal _MockClients.
-        return self._build_data_vars_clients(
-            structure, coords, optimize_wide_table=False
-        )
-
-    @property
-    def coords(self):
-        structure = self.structure().macro
-        coords_clients, _coords = self._build_coords_clients(structure)
-        return coords_clients
-
-    def _build_data_vars_clients(
-        self, structure, coords, variables=None, optimize_wide_table=True
-    ):
-        data_vars_clients = {}
-        wide_table_fetcher = _WideTableFetcher(
-            self.context.get_content, self.item["links"]["full_dataset"], coords
+    def _build_arrays(self, variables, optimize_wide_table):
+        data_vars = {}
+        coords = {}
+        # Optimization: Download scalar columns in batch as DataFrame.
+        # on first access.
+        coords_fetcher = _WideTableFetcher(
+            self.context.get_content, self.item["links"]["full"]
         )
-        for name, data_array in structure.data_vars.items():
+        if self.separate_coords_and_data_vars_download:
+            data_vars_fetcher = _WideTableFetcher(
+                self.context.get_content, self.item["links"]["full"]
+            )
+        else:
+            data_vars_fetcher = coords_fetcher
+        for name, array_client in self.items():
             if (variables is not None) and (name not in variables):
                 continue
-
-            metadata = self.metadata["data_vars"][name]
-            # Optimization: Download scalar data as DataFrame.
-            data_shape = data_array.macro.variable.macro.shape
-            if (
-                optimize_wide_table
-                and (data_shape[0] < LENGTH_LIMIT_FOR_WIDE_TABLE_OPTIMIZATION)
-                and (len(data_shape) < 2)
-            ):
-                data_vars_clients[name] = wide_table_fetcher.register(
-                    name, data_array, metadata
+            array_structure = array_client.structure()
+            shape = array_structure.macro.shape
+            if optimize_wide_table and (
```

Comment:
Wide table optimization also fails for "jagged" datasets but this is currently not caught by the server.
Consider: `dset = array.Dataset({f"{i}" :  xarray.DataArray(i*numpy.ones(2*i), dims=f"dim{i}") for i in range(10)})`

Reply:
I guess the wide table optimization should only be applied per dimension or more conservatively only if the dataset only has one dimension so it really is just a dataframe which can be transmitted using arrow.

Reply:
Good call: I had forgotten it was even possible to construct a ragged Dataset because the Bluesky Datasets always share a time coordinate and cannot be ragged.

This optimization really helps in cases where some (most) of the columns are scalar but some are ND.